derive/src/field_attributes.rs

```diff
@@ -0,0 +1,110 @@
+use proc_macro2::{Literal, TokenStream, TokenTree};
+use quote::quote;
+use syn::*;
+
+/// Used to filter out necessary field attribute and in panics.
+static ARBITRARY_ATTRIBUTE_NAME: &str = "arbitrary";
+
+/// Determines how a value for a field should be constructed.
+#[cfg_attr(test, derive(Debug))]
+pub enum FieldConstructor {
+    /// Assume that Arbitrary is defined for the type of this field and use it (default)
+    Arbitrary,
+
+    /// Places `Default::default()` as a field value.
+    Default,
+
+    /// Use custom function to generate a value for a field.
+    WithFunction(TokenStream),
+
+    /// Set a field always to the given value.
+    Value(TokenStream),
+}
+
+pub fn determine_field_constructor(field: &Field) -> FieldConstructor {
+    let opt_attr = fetch_attr_from_field(field);
+    match opt_attr {
+        Some(attr) => parse_attribute(attr),
+        None => FieldConstructor::Arbitrary,
+    }
+}
+
+fn fetch_attr_from_field(field: &Field) -> Option<&Attribute> {
+    field.attrs.iter().find(|a| {
+        let path = &a.path;
+        let name = quote!(#path).to_string();
+        name == ARBITRARY_ATTRIBUTE_NAME
+    })
+}
+
+fn parse_attribute(attr: &Attribute) -> FieldConstructor {
+    let group = {
+        let mut tokens_iter = attr.clone().tokens.into_iter();
+        let token = tokens_iter
+            .next()
+            .unwrap_or_else(|| panic!("{ARBITRARY_ATTRIBUTE_NAME} attribute cannot be empty."));
+        match token {
+            TokenTree::Group(g) => g,
+            t => panic!("{ARBITRARY_ATTRIBUTE_NAME} must contain a group, got: {t})"),
+        }
+    };
+    parse_attribute_internals(group.stream())
+}
+
+fn parse_attribute_internals(stream: TokenStream) -> FieldConstructor {
+    let mut tokens_iter = stream.into_iter();
+    let token = tokens_iter
+        .next()
+        .unwrap_or_else(|| panic!("{ARBITRARY_ATTRIBUTE_NAME} attribute cannot be empty."));
+    match token.to_string().as_ref() {
+        "default" => FieldConstructor::Default,
+        "with" => {
+            let func_path = parse_assigned_value(tokens_iter);
+            FieldConstructor::WithFunction(func_path)
+        }
+        "value" => {
+            let value = parse_assigned_value(tokens_iter);
+            FieldConstructor::Value(value)
+        }
+        _ => panic!("Unknown options for {ARBITRARY_ATTRIBUTE_NAME}: {token}"),
```

Comment:
Rather than panicking, can we return actual `Result`s and then in the main derive entry point and then either use https://docs.rs/syn/latest/syn/parse/struct.Error.html#method.to_compile_error or define our own `impl ToTokens for MyError` equivalent that expands into

```rust
quote! { compile_error!(#the_error_string) }
```

I think there isn't any reason we can't use `syn::Error` though.

This will provide better diagnostics for users.

Reply:
Thanks. I was not aware of `syn::Error`, a cool thing, I will added it to my toolbox.

Addressed within the following commits:
- https://github.com/rust-fuzz/arbitrary/pull/129/commits/f3225dddfead7034441829846394b6b30604f4c8 - refactor functions to return `syn::Result<TokenStream>`
- https://github.com/rust-fuzz/arbitrary/pull/129/commits/fb58d7a992bbcb42ebaedb0f23b2de435d41809e, https://github.com/rust-fuzz/arbitrary/pull/129/commits/8cbad14fd9c672413f15d3d53d1b2072cf72fbc9, https://github.com/rust-fuzz/arbitrary/pull/129/commits/479da322a5b65917bb933dc43b03a83ecdec0c03 - get rid of `panic!()`, return `syn::Error`.

This indeed resulted into better error messages :heart: . You can see the difference in `.stderr` files.